src/axom/lumberjack/tests/lumberjack_NonCollectiveRootCommunicator.hpp

```diff
@@ -0,0 +1,154 @@
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "axom/lumberjack/NonCollectiveRootCommunicator.hpp"
+#include "gtest/gtest.h"
+#include "mpi.h"
+
+TEST(lumberjack_NonCollectiveRootCommunicator, noncollective_communication)
+{
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  int commSize = -1;
+  MPI_Comm_size(MPI_COMM_WORLD, &commSize);
+
+  const int ranksLimit = 5;
+  axom::lumberjack::NonCollectiveRootCommunicator c;
+
+  c.initialize(MPI_COMM_WORLD, ranksLimit);
+
+  std::string message = std::to_string(c.rank());
+
+  std::vector<const char*> receivedPackedMessages;
+
+  // send message only from even ranks that are non-zero
+  if((c.rank() % 2) == 0 && c.rank() != 0)
+  {
+    c.push(message.c_str(), receivedPackedMessages);
+  }
+
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  // receive messages from rank 0 after barrier
+  if(c.rank() == 0)
+  {
+    c.push(message.c_str(), receivedPackedMessages);
+  }
+
+  if(c.rank() != 0)
+  {
+    EXPECT_EQ((int)receivedPackedMessages.size(), 0);
+  }
+  else
+  {
+    const int numMessagesToReceive =
+      ((commSize % 2) == 0) ? ((commSize / 2) - 1) : (commSize / 2);
+    EXPECT_EQ((int)receivedPackedMessages.size(), numMessagesToReceive);
+    for(int i = 1; i <= numMessagesToReceive; ++i)
+    {
+      std::string currMessage = std::to_string(i * 2);
+      bool found = false;
+      for(auto& rm : receivedPackedMessages)
+      {
+        if(strcmp(rm, currMessage.c_str()) == 0)
+        {
+          found = true;
+        }
+      }
+      EXPECT_EQ(found, true)
+        << "Message not received: " << currMessage << std::endl;
+    }
+  }
+
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  c.finalize();
+
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  // cleanup allocated memory from received messages
+  for(auto& rm : receivedPackedMessages)
+  {
+    delete[] rm;
+  }
+
+  receivedPackedMessages.clear();
+}
+
+TEST(lumberjack_NonCollectiveRootCommunicator, multiple_communicators)
+{
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  int commSize = -1;
+  MPI_Comm_size(MPI_COMM_WORLD, &commSize);
+
+  const int ranksLimit = 5;
+  axom::lumberjack::NonCollectiveRootCommunicator c1;
+  axom::lumberjack::NonCollectiveRootCommunicator c2;
+
+  c1.initialize(MPI_COMM_WORLD, ranksLimit);
+  c2.initialize(MPI_COMM_WORLD, ranksLimit);
+
+  std::vector<const char*> receivedPackedMessages_c1;
+  std::vector<const char*> receivedPackedMessages_c2;
+  const std::string c1_message = "c1";
+  const std::string c2_message = "c2";
+
+  if(c1.rank() == 1)
+  {
+    c1.push(c1_message.c_str(), receivedPackedMessages_c1);
+  }
+
+  if(c2.rank() == 1)
+  {
+    c2.push(c2_message.c_str(), receivedPackedMessages_c2);
+  }
+
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  /*
+    Sending/receiving messages from c1 should not 
+    interfere with sending/receiving from c2.
+    This tests that ordering of sends should not affect 
+    the order of messages being received.
+  */
+
+  if(commSize > 1 && c1.rank() == 0)
+  {
+    c1.push(nullptr, receivedPackedMessages_c1);
```

Comment:
> Update on the failing tests: The multiple_communicators test I added sporadically fails

@gberg617 I was able to fix the sporadic failure locally in my Docker container by doing the following where the root rank expects to receive messages:
```suggestion
    while(receivedPackedMessages_c1.size() == 0)
    {
      c1.push(nullptr, receivedPackedMessages_c1);
    }

```
To explain why I think the test was failing the check (this doesn't explain why the application terminates, still don't understand that):

- A non-blocking send is sent by rank 1. We don't know for sure when this completes.
- The ranks synchronize at a `MPI_Barrier`. **Notably**, this _does not guarantee_ that the non-blocking send has completed at the exit of the barrier.
- Rank 0 attempts to receive the non-blocking send, the send has not completed, and size 0 is returned, failing the check.

The added while loop lets the root rank poll until it sees the single message.

I think the reason this error didn't pop up on LC systems is because the inter and intra connects are so much faster than on a standard laptop/CI node.